### PR TITLE
Ksagiyam/add periodic hex mesh

### PR DESCRIPTION
**Pull request by ksagiyam**
Add periodic hex box mesh.

~Depends on https://gitlab.com/petsc/petsc/-/merge_requests/7677.~ Merged.

- Use `periodic` + `sparseLocalize=False` in `PETSc.DMPlex().createBoxMesh()`  (Firedrake requires DG coordinates on the entire domain),
- Use `RelabeledMesh` for marking boundaries.

### CODE REVIEW

**Comment by juliusgh** [comment]
Hi @ksagiyam, great that you are working on this!

I was just looking for this feature and saw your PR. In the meantime, the corresponding PETsc PR https://gitlab.com/petsc/petsc/-/merge_requests/7677 has been merged.

Do you know what is missing in your PR and do you intend to continue working on it? Otherwise, I could also take a look.

**Comment by ksagiyam** [comment]
I need to make this branch up-to-date. I also noticed that I had not labeled boundaries correctly, so I need to fix that. I will work on this later today.

File: firedrake/utility_meshes.py [hunk 1]
```diff
@@ -1809,112 +1812,187 @@ def PeriodicBoxMesh(
...
+    nz : int
+        Number of cells in the z direction.
+    Lx : float
+        Extent in the x direction.
+    Ly : float
+        Extent in the y direction.
+    Lz : float
+        Extent in the z direction.
+    directions : list or tuple or None
+        Directions of periodicity; if `None`, the mesh is made periodic in all directions.
```
[the remark this answers is not in the document]
Comment (ksagiyam) [reply firedrake/utility_meshes.py]: Yes. Fixed.

File: firedrake/utility_meshes.py [hunk 2]
```diff
@@ -1809,112 +1812,187 @@ def PeriodicBoxMesh(
...
+            [v0, v6, v4, v7],
+            [v0, v2, v6, v7],
+        ]
+        cells = np.asarray(cells).reshape(-1, ny, nx, nz).swapaxes(0, 3).reshape(-1, 4)
+        plex = mesh.plex_from_cell_list(
+            3, cells, coords, comm, mesh._generate_default_mesh_topology_name(name)
+        )
+        m = mesh.Mesh(
+            plex,
+            reorder=reorder_noop,
```
[the remark this answers is not in the document]
Comment (ksagiyam) [reply firedrake/utility_meshes.py:1944]: We only need correct reordering and distribution for the mesh returned by `_postprocess_periodic_mesh()`, so we just avoid unnecessary reordering and adding overlap when making this intermediate mesh `m`.

File: firedrake/utility_meshes.py [hunk 3]
```diff
@@ -1809,112 +1812,187 @@ def PeriodicBoxMesh(
...
+                = extent
 
-    return _postprocess_periodic_mesh(new_coordinates,
-                                      comm,
-                                      distribution_parameters,
-                                      reorder,
-                                      name,
-                                      distribution_name,
-                                      permutation_name)
+        return _postprocess_periodic_mesh(new_coordinates,
```
[the remark this answers is not in the document]
Comment (ksagiyam) [reply firedrake/utility_meshes.py:1978]: We call `_postprocess_periodic_mesh()` to make DMPlex DG coordinates from Firedrake DG coordinates. For periodic hex mesh, `createBoxMesh()` directly makes DMPlex DG coordinates (and we later convert these to Firedrake DG coordinates), so we do not need this additional step.

[the remark this answers is not in the document]
Comment (connorjward) [reply firedrake/utility_meshes.py:1978]: Got it. Thanks

File: firedrake/utility_meshes.py [hunk 4]
```diff
@@ -1809,112 +1812,185 @@ def PeriodicBoxMesh(
...
     """
     for n in (nx, ny, nz):
         if n < 3:
             raise ValueError(
                 "3D periodic meshes with fewer than 3 cells are not currently supported"
             )
+    if hexahedral:
+        if len(directions) != 3:
+            raise ValueError(f"directions must have exactly dim (=3) elements : Got {directions}")
+        plex = PETSc.DMPlex().createBoxMesh((nx, ny, nz), lower=(0., 0., 0.), upper=(Lx, Ly, Lz), simplex=False, periodic=directions, interpolate=True, sparseLocalize=False, comm=comm)
```
Comment (connorjward) [inline firedrake/utility_meshes.py:1876]:
```suggestion
        plex = PETSc.DMPlex().createBoxMesh(
            (nx, ny, nz),
            lower=(0., 0., 0.),
            upper=(Lx, Ly, Lz),
            simplex=False,
            periodic=directions,
            interpolate=True,
            sparseLocalize=False,
            comm=comm,
        )
```

**Review by connorjward — CHANGES_REQUESTED** [review]
One trivial thing. With that please consider this as approved by me.

**Comment by ksagiyam** [comment]
Thanks, Connor.

**Pull request merged**